src/server/db_slice.cc

```diff
@@ -1355,4 +1297,85 @@ void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice&
   }
 }
 
+void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
+  auto it = client_tracking_map_.find(key);
+  if (it != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = it->second;
+    auto cb = [key, client_set = std::move(client_set)](unsigned idx, util::ProactorBase*) {
+      for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+        if ((unsigned int)it->Thread() != idx)
+          continue;
+        facade::Connection* conn = it->Get();
+        if ((conn != nullptr) && conn->IsTrackingOn())
+          conn->SendInvalidationMessageAsync({key});
+      }
+    };
+    shard_set->pool()->DispatchBrief(std::move(cb));
+    // remove this key from the tracking table as the key no longer exists
+    client_tracking_map_.erase(key);
+  }
+}
+
+void DbSlice::PerformDeletion(PrimeIterator del_it, ExpireIterator exp_it, EngineShard* shard,
```

Comment:
Its hard to review with github. Have you done any changes in the function instead of moving it to be member of the class?

Reply:
no change in the function, but added invalidation notification at the end of the PerformDeletion implementation.